### DISCUSSION TITLE
Possible corner case bug: RHF validations halting chain of events after they run

### DISCUSSION THREAD

**Discussion by jamesfong** [react-hook-form/react-hook-form#7822] [Q&A] [answered]
Codesandbox here: https://codesandbox.io/s/custom-validation-issue-cy2ff?file=/src/App.js

Simple form that is extracted from a more complex form I'm working on.

- Last name has a custom validation that checks to see if Last Name is empty and First Name is set
- Submit button does a `trigger` and also prints to console
- Most clicking around, filling in the form, then clicking on the button leads to the expected output in console
- I don't get a print to console when I:

1. Type in a first name
2. Mouse click into last name input
3. Type in a last name
4. Delete last name
5. Mouse to and click "Submit" **_without clicking anywhere else_**. The error message shows up, but I do not get the console log saying the button was clicked. It's almost like the custom validation is preventing further events from happening. If you make a mouse click somewhere before hitting submit, it seems to work. Also, if you remove the custom validation, it works fine.

Any ideas?

**Comment by Moshyfawn** [comment, marked as answer]
What I think is happening is when you go to click your Submit button _(push down your mouse button over it)_, the `mode: onBlur` validation kicks in as you've just blurred _(unfocused)_ your input. That results in the layout shift, the `mouseleave` event fires and the button is no longer clicked. So, your `trigger` form method isn't executed along with your `console.log`. If it did, you'd get an additional re-render and the "Errors" console.log with form errors.

See this screen-cast with your desired behavior if no layout has shifted due to the field error pushing the button down.  

https://user-images.githubusercontent.com/16290753/154025037-d71a211d-4c8f-490e-a64b-1c3254583ada.mp4

**Reply by jamesfong** [reply]
Thanks for the reply.

In the screen casted example that's one that works correctly. I've tried many combinations and the one that I wrote out in the original post is the one that doesn't seem to work correctly. It's almost as if the custom validation fails in the sequence of events causing later events down the chain to halt.

If you blur off the custom validation field before clicking the button, the proper behavior is exhibited.

**Reply by jamesfong** [reply]
After some more testing, it looks like it's not specifically related to the custom validation. I'm seeing this happen with a simple `required: true` check.

The way to replicate this is:

- First is required, Last is required if first is set
- Click in first, then click out to populate the required error
- Fill in second
- Fill in first
- Mouse directly to the submit button and click. No console.log will show up showing the button was clicked.

There are a few variations of the above that don't work. I'm stumped.

**Reply by Moshyfawn** [reply]
So, I've tried a couple of different ways to submit the form with `onBlur` validation and the only thing I see is that the issue is the layout shift due to the `onBlur` validation firing right when you **press** _(not click)_ your mouse button over the submit button. I feel like it's the issue you're experiencing.

Take this screen-cast for example

https://user-images.githubusercontent.com/16290753/154125874-82cbfef4-b8cf-4172-ad68-1ef03e4e5452.mp4

Notice how pressing _(not yet clicking , releasing)_ the button occurs and the `onBlur` validation kicks in. _(In this screen-cast a circle around the cursor means "pressed" state)_ Before you have a chance to release the mouse button over the submit button, the layout has already shifted because of the errors element disappearing and releasing _(finishing you click)_ does nothing, because the **click** event happens if you release the mouse button **over** the submit button.

Now, if you drop your error messages below the submit button, the buttons position won't be affected and you gonna get your `console.log`

https://user-images.githubusercontent.com/16290753/154127146-021c1a66-2ab1-45b1-9140-4b50e71b4eb1.mp4

On top of that, if you hover your mouse over the submit button while holding the mouse button and release it, you will still get your click event

https://user-images.githubusercontent.com/16290753/154127648-6614e033-be67-42b6-9f07-e9b77a82c4bc.mp4

At the end of the day, the sequence of events is:
1. You press the mouse button over the submit button
2. The `onBlur` validation kicks in as you unfocus _(blur)_ your input by pressing the submit button
3. The validation results in an error that is displayed above the submit button which causes it to shift down
4. You release the mouse button but not over the submit button as it's shifted down because of step 3. It doesn't matter how fast you do it, the mouse button release event is gonna happen outside the submit button element, so no click event fired

**Reply by jamesfong** [reply]
WOW, I think that's it! Thanks so much for that extensive debug. It just so happens that the UI is the same in both my project and what I reworked in codesandbox.

Thanks again!

**Reply by Moshyfawn** [reply]
Totally understand! Glad to help!
